engine/pkg/errctx/ctx.go

```diff
@@ -20,41 +20,45 @@ import (
 
 type errCtx struct {
 	context.Context
-	center *ErrCenter
 
-	once   sync.Once
-	doneCh <-chan struct{}
+	mu     sync.Mutex
+	cancel context.CancelFunc
+	err    error
 }
 
-func newErrCtx(parent context.Context, center *ErrCenter) *errCtx {
+func newErrCtx(parent context.Context) *errCtx {
+	ctx, cancel := context.WithCancel(parent)
 	return &errCtx{
-		Context: parent,
-		center:  center,
+		Context: ctx,
+		cancel:  cancel,
 	}
 }
 
-func (c *errCtx) Done() <-chan struct{} {
-	c.once.Do(func() {
-		doneCh := make(chan struct{})
-
-		go func() {
-			select {
-			case <-c.center.doneCh:
-			case <-c.Context.Done():
-			}
-
-			close(doneCh)
-		}()
+func (c *errCtx) doCancel(err error) {
+	if err == nil {
+		panic("errctx: internal error: missing cancel error")
+	}
 
-		c.doneCh = doneCh
-	})
-	return c.doneCh
+	c.mu.Lock()
+	if c.err != nil {
+		c.mu.Unlock()
+		return // already canceled
+	}
+	if c.Context.Err() != nil {
+		c.err = c.Context.Err()
+		c.mu.Unlock()
+		return // parent is canceled
+	}
```

Comment:
Better to add a unit test to cover this logic.
Besides is it necessary to record parent context error instead of the passing in error?

Reply:
> // If Done is not yet closed, Err returns nil.
	// If Done is closed, Err returns a non-nil error explaining why:
	// Canceled if the context was canceled
	// or DeadlineExceeded if the context's deadline passed.
	// After Err returns a non-nil error, successive calls to Err return the same error.
	Err() error

`Err()` shouldn't return a different error if it has returned the parent context error.